clients/client/src/main/java/org/projectnessie/client/auth/NessieAuthenticationProvider.java

```diff
@@ -0,0 +1,105 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.client.auth;
+
+import static org.projectnessie.client.NessieConfigConstants.CONF_NESSIE_AUTH_TYPE;
+
+import java.util.Locale;
+import java.util.ServiceLoader;
+import java.util.function.Function;
+
+/**
+ * Base interface for different authentication methods like "basic" (username + plain password),
+ * bearer token, etc.
+ */
+public interface NessieAuthenticationProvider {
+
+  /**
+   * Configure a {@link NessieAuthentication} instance using the configuration supplied by {@code
+   * configuration}.
+   *
+   * <p>Checks all instances returned by Java's {@link ServiceLoader} for the {@link
+   * NessieAuthenticationProvider} interface.
+   *
+   * <p>Uses the config option {@link
+   * org.projectnessie.client.NessieConfigConstants#CONF_NESSIE_AUTH_TYPE} to determine the
+   * authentication type by comparing the config value with the value returned by implementations'
+   * {@link #getAuthTypeValue()}. If {@link
+   * org.projectnessie.client.NessieConfigConstants#CONF_NESSIE_AUTH_TYPE} is not configured, {@code
+   * null} will be returned. If no implementation could be found, an {@link
+   * IllegalArgumentException} is thrown.
+   *
+   * <p>If a {@link NessieAuthenticationProvider} instance was found, the implementation's builder
+   * {@link #build(Function)} will be called.
+   */
+  static NessieAuthentication fromConfig(Function<String, String> configuration) {
+    String authType = configuration.apply(CONF_NESSIE_AUTH_TYPE);
+    if (authType != null) {
+      for (NessieAuthenticationProvider ap :
+          ServiceLoader.load(NessieAuthenticationProvider.class)) {
+        if (ap.getAuthTypeValue().equalsIgnoreCase(authType)) {
+          return ap.build(configuration);
+        }
+      }
+
+      // Fallback to reflection, just in case the files in META-INF/services required by
+      // ServiceLoader got lost by broken shaded jars.
+
+      String reflectionClassName = NessieAuthentication.class.getName();
+      reflectionClassName =
+          reflectionClassName.substring(0, reflectionClassName.lastIndexOf('.') + 1)
+              + authType.substring(0, 1).toUpperCase(Locale.ROOT)
+              + authType.substring(1).toLowerCase(Locale.ROOT)
+              + "AuthenticationProvider";
+
+      NessieAuthenticationProvider ap = null;
+      try {
+        @SuppressWarnings("unchecked")
+        Class<NessieAuthenticationProvider> provider =
+            (Class<NessieAuthenticationProvider>) Class.forName(reflectionClassName);
+        ap = provider.getDeclaredConstructor().newInstance();
+      } catch (Exception ignore) {
+        // ignore
```

Comment:
Should we at least debug log it? Class loading problems can be very hard to debug without evidence :thinking:

Reply:
Here it could be loading problems, cast problem, clinit problems... WDYT?

Reply:
+1 its an awful broad exception